spring-data-mongodb/src/main/java/org/springframework/data/mongodb/repository/query/AbstractReactiveMongoQuery.java

```diff
@@ -415,17 +463,65 @@ protected Mono<ParameterBindingDocumentCodec> getParameterBindingCodec() {
 	 * @param accessor must not be {@literal null}.
 	 * @return a {@link Mono} emitting the {@link SpELExpressionEvaluator} when ready.
 	 * @since 3.4
+	 * @deprecated since 4.3, use
+	 *             {@link #getValueExpressionEvaluatorLater(ExpressionDependencies, MongoParameterAccessor)} instead
 	 */
+	@Deprecated(since = "4.3")
 	protected Mono<SpELExpressionEvaluator> getSpelEvaluatorFor(ExpressionDependencies dependencies,
 			MongoParameterAccessor accessor) {
-
-		return evaluationContextProvider
-				.getEvaluationContextLater(getQueryMethod().getParameters(), accessor.getValues(), dependencies)
-				.map(evaluationContext -> (SpELExpressionEvaluator) new DefaultSpELExpressionEvaluator(expressionParser,
-						evaluationContext))
+		return valueEvaluationContextProvider.getEvaluationContextLater(accessor.getValues(), dependencies)
+				.map(evaluationContext -> (SpELExpressionEvaluator) new DefaultSpELExpressionEvaluator(
+						new SpelExpressionParser(), evaluationContext.getEvaluationContext()))
 				.defaultIfEmpty(DefaultSpELExpressionEvaluator.unsupported());
 	}
 
+	/**
+	 * Obtain a {@link ValueExpressionEvaluator} suitable to evaluate expressions.
+	 *
+	 * @param accessor must not be {@literal null}.
+	 * @since 4.3
+	 */
+	ValueExpressionEvaluator getValueExpressionEvaluator(MongoParameterAccessor accessor) {
+
+		return new ValueExpressionEvaluator() {
+
+			@Override
+			public <T> T evaluate(String expressionString) {
+				ValueExpression expression = valueExpressionDelegate.parse(expressionString);
+				ValueEvaluationContext evaluationContext = valueEvaluationContextProvider.getEvaluationContext(accessor.getValues(),
+						expression.getExpressionDependencies());
+				return (T) expression.evaluate(evaluationContext);
+			}
+		};
+	}
+
+	/**
+	 * Obtain a {@link Mono publisher} emitting the {@link ValueExpressionEvaluator} suitable to evaluate expressions
+	 * backed by the given dependencies.
+	 *
+	 * @param dependencies must not be {@literal null}.
+	 * @param accessor must not be {@literal null}.
+	 * @return a {@link Mono} emitting the {@link ValueExpressionEvaluator} when ready.
+	 * @since 4.3
+	 */
+	protected Mono<ValueExpressionEvaluator> getValueExpressionEvaluatorLater(ExpressionDependencies dependencies,
+			MongoParameterAccessor accessor) {
+
+			return valueEvaluationContextProvider.getEvaluationContextLater(accessor.getValues(), dependencies)
+					.map(evaluationContext -> {
+
+						return new ValueExpressionEvaluator() {
+							@Override
+							public <T> T evaluate(String expressionString) {
+
+								ValueExpression expression = valueExpressionDelegate.parse(expressionString);
+
+								return (T) expression.evaluate(evaluationContext);
+							}
+						};
+					});
+	}
```

Comment:
can we move the inner `ValueExpressionEvaluator` to a dedicated class that holds the evaluation context, so we can use for both?